opencog/cython/PythonSCM.cc

```diff
@@ -97,21 +97,13 @@ void* PythonSCM::init_in_guile(void* self)
 {
 	scm_c_define_module("opencog python", init_in_module, self);
 	scm_c_use_module("opencog python");
-
-	// Make sure that guile and python are using the same atomspace.
-	// This will avoid assorted confusion.
-	AtomSpace* as = SchemeSmob::ss_get_env_as("python-eval");
-	PythonEval& pev = PythonEval::instance(as);
+	PythonEval& pev = PythonEval::instance();
 
 	// This feels hacky, I guess, but I cannot figure out any other
 	// way of telling python which atomspace it is supposed to use
 	// by default.
 	pev.eval("from opencog.atomspace import AtomSpace");
 	pev.eval("from opencog.type_constructors import set_type_ctor_atomspace");
-	char buf[301];
-	snprintf(buf, 300, "set_type_ctor_atomspace(AtomSpace(%ld))\n",
-		(uint64_t) as);
-	pev.eval(buf);
```

Comment:
Agree, that approach used in this code is error-prone, but removing this changes behavior. 
In scenario when Python code is being called from Scheme and Python is creating new atom using atom type constructor default atomspace is None and it should lead to errors. Nevertheless first example I found where Python is called from Scheme accurately uses atomspace in Python https://github.com/opencog/atomspace/blob/master/examples/atomspace/python.scm and isn't affected.

Reply:
One can set default atomspace for python from scheme calling initialize_opencog from python. Though i'll split initialize_opencog into two methods in another PR.

Reply:
After thinking about it I realized that as we have an API which implicitly sets an atomspace into atom constructors then initializing the atomspace when Scheme initializes Python is normal. This PR is about splitting Python API initialization and type constructor atomspace setter so these both steps should be used when initializing Python from Scheme.

The problem is that in the following two scenarios we should have the different behavior:
1. Scheme initializes Python 
2. Python initializes Scheme which initializes Python

In the first scenario we should set Scheme's atomspace as a type constructor atomspace for Python.
In the second scenario we should set Python's atomspace as a type constructor atomspace for Scheme and during the second initialization just do nothing. To do this we need to know whether PythonEval was initialized for the first time or not. If PythonEval is just initialized then we need to set type constructor atomspace.

Reply:
user can set atomspace from scheme using:
```scheme
(python-call-with-as "set_type_ctor_atomspace" (cog-atomspace))
```

Though I can make common global pointer to default atomspace for both python and scheme

Reply:
This creates a bug for the user ... it is not reasonable for the user to know that sometimes, under certain special conditions, that python and scheme sometimes use two different atomspaces, and then try to guess when this might happen, and try to manually patch that up in their code.  It's a unified system, it should behave like a unified system.

Reply:
There's also a third scenario:

3. Cogserver initializes AtomSpace, and then initializes Scheme and Python

Reply:
If we implicitly share default atomspace between interpreters, then what should happen if one of interpreters change default atomspace?

Reply:
Ahh, good question. Nothing, I think.  There's no "default" atomspace, not really; there's only the "initial atomspace created when system is started" and after that, its the "atomspace of the current context", the context being per-thread. There are four "common" contexts:

1. During pattern matching, partial results are evaluated in a temp atomspace; the correct atomspace to use is set using the arguments to the c++ `ExecutationOutputLink` and the c++ `EvaluationLink`. Sometimes, these can be two or three deep, because Gets/Puts can be nested arbitrarily. (Yes, this does happen).  Also, different threads can have different atomspaces (e.g. one thread is handling sensory input, e.g. from cameras; a different thread is doing PLN/mining/language/etc) 

2. During pattern mining ... there is also some kind of temp atomspace, I don't know what its for.  It's not one of the temps used by the matcher.

3. in agi-bio, the default atomspace is read-only and shared; its huge. All updates (TV changes, adds, deletes) occur in an overlay atomspace that is read-write.  (and pattern matching is then in a temp atomspace on top of this overlay) It's done this way to save disk space: the shared-read-only is huge and contains genetic data, everyone uses the same copy, and then everyone creates just minor patches to it, for different experiments.

4. in guile, at least, atomspaces are per-thread, so that two different threads can run at the same time, each thread using a different atomspace.  I'm not sure where this is used.

Reply:
Vitaly had great idea to use thread-local queue per thread.  I then we can set current default atomspace per thread instead of per interpreter. 

Somehow like this:
scheme or python call c++ method

```c++
push_atomspace_for_thread(std::shared_ptr<AtomSpace> a)
```

then scheme or python can call c++ factory for atom creation:
```c++
AtomPtr create_with_current_context(Type t, ...args)
```

where create_with_current_context takes atomspace from thread-local queue

Reply:
Though with this design one would have to set atomspace upon thread creation, hmm..

Reply:
To explain the idea further Anatoly and me discussed that working with two global variables on two platforms (Scheme and Python) is really difficult. Use cases described above suggest one need to set the global atomspace variable before calling Python from Scheme (or vice versa) and revert the value after returning from the call. And what if this global variable is changed by the code in the parallel thread.

We have concluded the following:
- each thread should have its own pointer to the current atomspace which is used by the code of the thread;
- each thread should have API to change its pointer before executing block of code and  revert this change after block is executed (something similar to `(cog-push-atomspace)` `(cog-pop-atomspace)` in Scheme);
- this pointer should be shared by Scheme, Python and C++ code.

Under these three conditions it doesn't matter how complex is a chain of the calls. C++ -> Python -> Scheme -> Python -> C++. The atomspace pointer is always same.

Reply:
@linas I see that in PythonSCM::init_in_guile there is code that imports in python set_type_ctor_atomspace. Comment says it supposed to set default atomspace. It looks like it doesn't, where the default atomspace should come from?

Reply:
I put this code back with some refactoring